Hoist the invariants out of multi-level nested loops

kunalspathak opened the pull request:
This is an experiment to hoist expressions out of multi-level loops. Currently, we just hoist expressions one-level up because of which some invariants are evaluated repeatedly in outer loop.

The shortcoming is because of two reasons from what I understand. First, we always hoist outer loop and then go and try to hoist code from inner loop as seen in this comment:

https://github.com/dotnet/runtime/blob/1a296c06fff8750b4658f6f7d901347e006744d0/src/coreclr/jit/optimizer.cpp#L6202-L6208

The second part is, we only evaluate the blocks that are immediate dominators of the exit blocks and do not evaluate conditional blocks that are frequently executed.

To overcome both of these problems, I changed the ordering of hoisting from inner to outer. Thus, we will first hoist all the nested loops present inside the current loop before hoisting the code present in the current loop. Now, once we create the preheader from the code hoisted from nested loop, we need to make sure that the preheader is considered while evaluating if it can be hoisted even outside the parent loop. For that, I keep track of newly created preheaders and then add them in the list of blocks to be evaluated for hoisting. When the hoisting is checked for the parent loop, if the code inside the preheader (that was added for the invariant code present in nested loop) is also invariant, it will get hoisted outside the parent loop too. 

I see nice diffs from this change:

Example 1 taken from https://github.com/dotnet/runtime/issues/35735#issuecomment-630410230:

![image](https://user-images.githubusercontent.com/12488060/163490945-f24a90a1-3004-49cd-907f-f52b3dbb206b.png)

Example 2 taken from https://github.com/dotnet/runtime/issues/61420#issuecomment-1029696439:

![image](https://user-images.githubusercontent.com/12488060/163490975-479dc4a3-6607-4a72-9afa-e91f2c4f583d.png)

Example 3 taken from https://github.com/dotnet/runtime/issues/61420#issuecomment-965450829

![image](https://user-images.githubusercontent.com/12488060/163529001-af768169-bcc6-42ca-9eb6-942f975387c1.png)

Example 4 taken from Imagesharp code in https://github.com/dotnet/runtime/issues/61420#issuecomment-965617066

![image](https://user-images.githubusercontent.com/12488060/168456629-6faf1d04-2cf3-43fa-b551-bf618b56de52.png)

Example 5 taken from https://github.com/dotnet/runtime/issues/6666#issuecomment-258723866

![image](https://user-images.githubusercontent.com/12488060/163529836-d0c737e6-a96c-4817-91d0-80dedfde27d7.png)

Contributes to https://github.com/dotnet/runtime/issues/65342


TODO:
- [x] With the inner->outer flow, revisit `GetHoistedInCurLoop()` and `ExtractHoistedInCurLoop()`
- [x] The immediate problem of hoisting past an expression that can throw is solved. The recent commit a4772a08a11 handles it by keeping track of `GT_COMMA` nodes and see if any of the siblings that occur before the node that we want to hoist (in the program execution) throws exception. If yes, it will prohibit the node from hoisting. However, I noticed that there are few cases where today, we "aggressively hoist" node from the `GT_COMMA` node even if the earlier siblings (nodes that will execute before the loop in question) might throw exception. I need to check if the existing mechanism is correct and if yes, adjust my conditions added in a4772a08a11  to adapt to it, so we don't see any regressions.
- [x] Are there any other conditions that we should consider to prohibit from "aggressive hoisting" that might introduce functional correctness issues? Verify with some counter examples.
- [x] Currently the preheader of inner loops are visited last because I am using stack to gather them and we pop them in reverse order. Fix it so the preheaders are processed in execution order.

kunalspathak committed:
first working version

kunalspathak committed:
Skip check of VN hoisting

kunalspathak commented:
@dotnet/jit-contrib , @BruceForstall , @AndyAyersMS - any feedback on the approach that I am taking here?

AndyAyersMS commented:
> @dotnet/jit-contrib , @BruceForstall , @AndyAyersMS - any feedback on the approach that I am taking here?

So you are enabling conditional hoisting but just from inner loop preheaders? Seems like if we are going to do this we ought to think more about enabling conditional hoisting in general. But this will put pressure on CSE/LSRA.

FWIW in previous compilers we would consider conditional hoisting based on profile data if the loop block execution frequency was an appreciable fraction of the entry block frequency. This will "work" in either inner->outer or outer->inner hosting approaches, and it may even be that outer->inner works better overall as even a thinly executed inner loop block may end up getting hoisted as the benefit gets multiplied the further you can hoist it.

I don't know if our profile maintenance gives us reasonable numbers for the preheader block weight. There are a few places that are involved -- first loop inversion has to impute some kind of zero-trip likelihood; second preheader creation has to impute the right weight for the preheader (which will in general not be the weight of either its predecessor or successor).

I also don't know if we do the right safety checks to allow conditional hoisting; I suspect we don't.

kunalspathak commented:
> So you are enabling conditional hoisting but just from inner loop preheaders? Seems like if we are going to do this we ought to think more about enabling conditional hoisting in general. But this will put pressure on CSE/LSRA.

Yes, my idea is to start with smaller scenarios and get them in for .NET 7. As you see in the examples, it catches the common cases (specifically things like hoisting the static variables addresses for arm64). In future releases, we can build more general conditional hoisting.

> But this will put pressure on CSE/LSRA.

If I can careful hoist the frequently executing loop blocks, it might still prove beneficial in execution time and might cover the register pressure impact.

kunalspathak commented:
@BruceForstall , to answer your question offline for the `mul` example above, the reason is pretty much what I was mentioning. When we check the things to be hoisted from outside the loop, we don't consider the block that contains the multiplication operation. Here are the blocks before loop hoisting happens with following loops:

```
L00, from BB02 to BB04 (Head=BB01, Entry=BB02, Exit=BB04), child loop = L01
L01, from BB03 to BB03 (Head=BB02, Entry=BB03, Exit=BB03, parent=L00)
```

```
-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..006)-> BB05 ( cond )                     i 
BB02 [0001]  2       BB01,BB04             4     0 [006..00A)-> BB04 ( cond )                     i Loop Loop0 bwd bwd-target 
BB03 [0002]  2       BB02,BB03            16     1 [00A..019)-> BB03 ( cond )                     i Loop Loop0 bwd bwd-target align 
BB04 [0004]  2       BB02,BB03             4     0 [019..022)-> BB02 ( cond )                     i bwd 
BB05 [0006]  2       BB01,BB04             1       [022..024)        (return)                     i 
-----------------------------------------------------------------------------------------------------------------------------------------
```

We first try to hoist `L00` and only consider blocks `BB02` and `BB04` while the `mul` operation is inside `BB03`.

```
Loop Nest L00
optHoistThisLoop for loop L00 <BB02..BB04>:
  Loop body does not contain a call
  Loop has single exit

  USEDEF  (4)={V03 V00 V01 V02}
  INOUT   (4)={V03 V00 V01 V02}
  LOOPVARS(4)={V03 V00 V01 V02}
  Only considering hoisting in blocks that dominate exit block BB04

    optHoistLoopBlocks BB02 (weight=  4   ) of loop L00 <BB02..BB04>
      [000012] not hoistable: variant child
      [000040] not hoistable: variant child

    optHoistLoopBlocks BB04 (weight=  4   ) of loop L00 <BB02..BB04>
      [000033] not hoistable: variant child
      [000009] not hoistable: variant child
```

Next, we inspect `L01` and decide to hoist the operation from inside that loop to outside, but now it is too late to be hoisted further out.

```
Nested Loop L01
optHoistThisLoop for loop L01 <BB03..BB03>:
  Loop body does not contain a call
  Loop has single exit

  USEDEF  (3)={V03 V00 V01}
  INOUT   (4)={V03 V00 V01 V02}
  LOOPVARS(3)={V03 V00 V01}
  Only considering hoisting in blocks that dominate exit block BB03

    optHoistLoopBlocks BB03 (weight= 16   ) of loop L01 <BB03..BB03>
      [000023] not hoistable: unset
```

kunalspathak committed:
Account for duplicate blocks

kunalspathak committed:
clean up

kunalspathak commented:
/azp run runtime-coreclr outerloop

kunalspathak commented:
The outerloop surprisingly is green, but I did find an example where we would AV because of eager hoisting.

```c#
[MethodImpl(MethodImplOptions.NoInlining)]
static int Problem1_Av(bool c, int z, int[] a)
{
    int r = 0;
    for (int i = 0; i < 10; i++)
    {
        if (c)
        {
            for (int j = 0; j < 10; j++)
            {
                r += a[z];
            }
        }
    }
    return r;
}
```

Here, we hoist `a[z]` outside the outer-loop while the bounds check stays outside the inner-loop leading the AV.

```asm
G_M50606_IG01:
       push     rsi
       sub      rsp, 32
                                                ;; size=5 bbWeight=0    PerfScore 0.00
G_M50606_IG02:
       xor      eax, eax
       xor      r9d, r9d
       movzx    rcx, cl
       mov      r10d, edx
       mov      r10d, dword ptr [r8+4*r10+16]  ; <-- AV
                                                ;; size=16 bbWeight=1    PerfScore 3.00
G_M50606_IG03:
       test     ecx, ecx
       je       SHORT G_M50606_IG06
                                                ;; size=4 bbWeight=4    PerfScore 5.00
G_M50606_IG04:
       xor      r11d, r11d
       mov      esi, dword ptr [r8+8]
       cmp      edx, esi
       jae      SHORT G_M50606_IG08
       mov      esi, r10d
       align    [0 bytes for IG05]
                                                ;; size=14 bbWeight=2    PerfScore 7.50
G_M50606_IG05:
       add      eax, esi
       inc      r11d
       cmp      r11d, 10
       jl       SHORT G_M50606_IG05
                                                ;; size=11 bbWeight=16    PerfScore 28.00
G_M50606_IG06:
       inc      r9d
       cmp      r9d, 10
       jl       SHORT G_M50606_IG03
                                                ;; size=9 bbWeight=4    PerfScore 6.00
G_M50606_IG07:
       add      rsp, 32
       pop      rsi
       ret
                                                ;; size=6 bbWeight=1    PerfScore 1.75
G_M50606_IG08:
       call     CORINFO_HELP_RNGCHKFAIL
       int3

```
```

AndyAyersMS commented:
> > I also don't know if we do the right safety checks to allow conditional hoisting; I suspect we don't.

> I did find an example where we would AV because of eager hoisting.

Thanks for looking into this.

kunalspathak committed:
wip

kunalspathak committed:
isCommaTree && hasExcep

kunalspathak committed:
revert lsra changes

kunalspathak commented:
Failures are https://github.com/dotnet/runtime/issues/67816 and https://github.com/dotnet/arcade/issues/9208

kunalspathak committed:
Update hoisting condition

- Only update if node to be hoisted has side-effects and the sibling that is before that throws exception

kunalspathak committed:
Change to BasicBlockList

kunalspathak committed:
organize preheaders

kunalspathak committed:
update hoistedInCurLoop and hoistedInSiblingLoop

kunalspathak committed:
Reverse the loop order

kunalspathak committed:
cleanup and jit-format

kunalspathak committed:
Revert "Minor fix to display IG01 weight correctly"

This reverts commit 757120e863b2da188db2593da1b7142fd1ecf191.

kunalspathak commented:
/azp run runtime-coreclr outerloop, Fuzzlyn, Antigen

kunalspathak commented:
Antigen and Fuzzlyn failures are unrelated

kunalspathak commented:
/azp run runtime-coreclr jitstress, runtime-coreclr libraries-jitstress

kunalspathak committed:
simplify code

kunalspathak commented:
/azp run runtime-coreclr jitstress, runtime-coreclr libraries-jitstress, runtime-coreclr outerloop

kunalspathak commented:
Failures are https://github.com/dotnet/runtime/issues/69154, https://github.com/dotnet/runtime/issues/68530

kunalspathak commented:
@dotnet/jit-contrib - This is ready for review.

kunalspathak commented:
From https://github.com/dotnet/runtime/issues/35735#issuecomment-630410230, `InvariantInBothDependent`:

```diff
 ; rsp based frame
 ; fully interruptible
 ; No PGO data
-; invoked as altjit
 ; Final local variable assignments
 ;
-;  V00 arg0         [V00,T03] (  5, 14   )     int  ->  rcx         single-def
+;  V00 arg0         [V00,T04] (  5,  5   )     int  ->  rcx         single-def
 ;  V01 loc0         [V01,T01] (  4, 34   )     int  ->  rax
-;  V02 loc1         [V02,T04] (  4, 13   )     int  ->  rdx
+;  V02 loc1         [V02,T03] (  4, 13   )     int  ->  rdx
 ;  V03 loc2         [V03,T00] (  4, 52   )     int  ->   r8
 ;* V04 loc3         [V04    ] (  0,  0   )     int  ->  zero-ref
 ;# V05 OutArgs      [V05    ] (  1,  1   )  lclBlk ( 0) [rsp+00H]   "OutgoingArgSpace"
-;  V06 cse0         [V06,T02] (  2, 20   )     int  ->   r9         "CSE - aggressive"
+;  V06 cse0         [V06,T02] (  2, 17   )     int  ->  rcx         "CSE - aggressive"
 ;
 ; Lcl frame size = 0

@@ -22,20 +21,20 @@ G_M26730_IG01:
 G_M26730_IG02:
        xor      eax, eax
        xor      edx, edx
-                                               ;; size=4 bbWeight=1    PerfScore 0.50
+       mov      r8d, ecx
+       imul     r8d, ecx
+       imul     ecx, r8d
+                                               ;; size=15 bbWeight=1    PerfScore 4.75
 G_M26730_IG03:
        xor      r8d, r8d
-       mov      r9d, ecx
-       imul     r9d, ecx
-       imul     r9d, ecx
        align    [0 bytes for IG04]
-                                               ;; size=14 bbWeight=4    PerfScore 18.00
+                                               ;; size=3 bbWeight=4    PerfScore 1.00
 G_M26730_IG04:
-       lea      eax, [r9+rax+1]
+       lea      eax, [rcx+rax+1]
        inc      r8d
        cmp      r8d, 10
        jl       SHORT G_M26730_IG04
-                                               ;; size=14 bbWeight=16    PerfScore 40.00
+                                               ;; size=13 bbWeight=16    PerfScore 40.00
 G_M26730_IG05:
        inc      edx
        cmp      edx, 10
@@ -45,6 +44,6 @@ G_M26730_IG06:
        ret
                                                ;; size=1 bbWeight=1    PerfScore 1.00

-; Total bytes of code 40, prolog size 0, PerfScore 69.50, instruction count 15, allocated bytes for code 40 (MethodHash=2ba89795) for method ConsoleApp15.loop_hoist:InvariantInBothDependent(int):int
+; Total bytes of code 39, prolog size 0, PerfScore 56.65, instruction count 15, allocated bytes for code 39 (MethodHash=2ba89795) for method ConsoleApp15.loop_hoist:InvariantInBothDependent(int):int
 ; ============================================================


```

kunalspathak commented:
Ping

AndyAyersMS commented:
I am a little confused by the `vnSetHoistedForSibling` logic. 

If a loop has two nested child loops, there's no guarantee that they have any particular flow relationship to one another. So, I don't understand why the hoisting we do in one child should impact the hoisting we do in the other.

For example, if we have
```C#
int x = ...;

for (int i = 0; i < N, i++)
{
     if (p)
     {
          for (int j0 = 0; j0 < N0; j0++)
          {
                k = i * i;
                m = x * x;
          }
     }
    else
     {
          for (int j1 = 0; j1 < N1; j1++)
          {
                k = i * i;
                m = x * x;
          }
     }
}
```
presumably we should be able to hoist `i * i` out of both loops, but perhaps not to a common def.

When there is a common invariant like `x * x` I get that we'd like to hoist it out of the parent just once, not twice, so it seems like tracking still has a purpose.

kunalspathak commented:
> Let me double check how many expressions are missed from getting hoisted

Not many are missed actually because we won't hoist 10+ expressions (very rare) from the loop. From the superpmi asmdiffs:

1. asp: 2 methods
2. benchmarks: 4 methods
3. libraries.pmi: 1 method
4. coreclr_tests; 5 methods

So I will remove the  `vnSetHoistedForSibling` check.

kunalspathak committed:
Remove m_hoistedVNInSiblingLoop

kunalspathak committed:
Add back ResetHoistedInCurLoop

Fix igWeight

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -6054,46 +6071,60 @@ void Compiler::optHoistLoopNest(unsigned lnum, LoopHoistContext* hoistCtxt)
+        BitVec       m_visited(BitVecOps::MakeEmpty(&m_visitedTraits));
 
+        // Since loops in optLoopTable are arranged in reverse execution order, arrange
+        // them back in execution order before processing.
```
This is misleading; there is no defined execution order for the child loops.

Also, why do we care about the ordering?

kunalspathak replied:
This was done as part of adding `vnSetHoistedForSibling`. Turns out that we don't need it anymore, I can well not do the reordering. 

But here are my notes just to show the problem. Imagine we have a method that has following loop hierarchy. 

```
L00   
   L01
   L02
      L03 
      L04
   L05
      L06
```
Earlier, we would iterate from outer to inner, so we would visit parent loop if there is anything to hoist and then look into child loops. I am changing it to first look into child loops and then look into parent loop. The reason being that if we hoist out something from inner loop, by looking in parent *later* we can try to hoist that expression further out of parent loop as well. So far good.I was trying to use a cache where I can check if an expression with same VN has been already hoisted by a sibling loop (loop at same level, eg. L01, L02 and L05 is one sibling group; L03,L04 is another), then skip hoisting it. So I would imagine, I would first inspect L01, hoist stuff out of it. When I inspect L02 I can check in cache if it was something already hoisted and not worry hoisting it. However, I discovered that we don't even iterate over the child loops in execution order. We iterate over them in reverse execution order. Basically with my change of inner to outer, it would be like this: 

```
 L05 
 L06
 L02
 L04
 L03
 L01
```

The downside of reverse order is, if I hoist something out of L05,  I can't use cache during `L01` processing if something is already hoisted by sibling loop because the hoisted expression follows after L01 in execution order. 

Ideally, the processing order should be:
 
```
 L01
 L03
 L04
 L02
 L06
 L05
```

~If I don't fix it, not using cache is an alternative, but turns out cache plays a big role in determining the profitability. Every time we hoist an expression, we could no. of expressions hoisted. When calculating profitability, we relate no. of expressions hoisted vs. no. of registers we have and might reject if we exceed number of expressions hoisted for a loop.~ Turns out this is not much benefit and it is very rare to have multiple similar expressions that are all present in nested loops and are getting hoisted.

kunalspathak replied:
When reverting the "reverse" flow, I noticed yet another example of https://github.com/dotnet/runtime/issues/66994.

Just because the preheader's `bbNum` are different, we see diffs.

![image](https://user-images.githubusercontent.com/12488060/172680652-19a4db5b-2306-4f8b-ae9f-b518bed1afd6.png)

https://www.diffchecker.com/WOEwzTNo

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -6205,42 +6236,96 @@ void Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt)
         // Push dominators, until we reach "entry" or exit the loop.
+        // Also push the preheaders that were added for the nested loops,
+        // if any, along the way such that the final block list is arranged
+        // in execution order
```
For the dominator-walk added blocks calling this "execution order" makes sense. 

But for the child preheaders, less so.

I think you are trying to say that the order of the blocks you end up with is consistent with a dominator tree preorder.

I wonder if we should just create preheaders up front before building dominators, so that we don't have to do all this extra work to figure out how they fit in.

kunalspathak replied:
> I wonder if we should just create preheaders up front before building dominators

Bruce tried to do it but ran into some problems. I will follow up with him to revamp that work, but if it gets too complicated, I could do it in a follow-up PR.

kunalspathak replied:
I tried to force create loop preHeaders, but there are lot of design problems that needs to be addressed. Currently, they are created only during loop hoisting and other loop optimizations like cloning, unrolling, etc. are not aware of them. Once we start creating preheaders, their state needs to be taken into account or updated when we do these loop optimizations. I was debugging a case where we create a loop preheader and then clone the loop. In such cases, we just update the loop head in `optUpdateLoopHead()` to newer head, but we do not take into account if it was marked as prehead. In such circumstances, we should ideally create a new loop preheader and unmark the existing preheader. 

![image](https://user-images.githubusercontent.com/12488060/172878106-8416f9e2-4647-4bf2-b7c5-e206cf3f1e8d.png)

There might be similar complexities involved while unrolling the loop. Not that none of this is not doable, but I think it is needs some design thinking and might not be worth doing as part of this PR. 

I also realized that if I do preheader, it changes the bbNum of blocks. And even if nothing else changes, it definitely impacts the LSRA block sequencing which triggers lot of diffs.

On the side note, in this PR, I am conservative where I check if a sibling of a node can throw exception and if yes, do not bother hoisting it. Sometimes this might negative impact us. For instance, in one of the example, we do not hoist the array access because of presence of bound check, but then later, we eliminate that very bound check through assertion prop, but we didn't have that information earlier. But from correctness perspective, we should continue to be conservative in these things.

kunalspathak replied:
Ah, didn't realize you suggested to rephrase. Done.

AndyAyersMS reviewed (commented):
I think we should consider creating the preheaders earlier, perhaps right before building SSA?

Then some of this accounting becomes simpler; you push the preheaders once they are dominated by the next block in the tree walk up from exit.

kunalspathak commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -6054,46 +6071,60 @@ void Compiler::optHoistLoopNest(unsigned lnum, LoopHoistContext* hoistCtxt)
-                hoistCtxt->m_hoistedInParentLoops.Set(keys.Get(), true);
-            }
-        }
+        BitVecTraits m_visitedTraits(fgBBNumMax * 2, this);
```
Yeah, this was little conservative. I can probably do something like `fgBBNumMax + MAX_LOOP_NUM`.

kunalspathak committed:
Remove reversal of loop processing order

kunalspathak committed:
jit format

kunalspathak committed:
Experimental: Also generate PerfScore:

AndyAyersMS commented on src/coreclr/scripts/superpmi_diffs.py:
```diff
@@ -207,6 +207,8 @@ def main(main_args):
         "asmdiffs",
         "--no_progress",
+        "--metrics", "CodeSize",
+        "--metrics", "PerfScore",
```
Are you doing this so we see PerfScore reports in the CI?

I am on the fence about this. I'd like to keep the default simple.

kunalspathak replied:
Yes, I will revert this change before merging. This is just to see the PerfScore impact for all the collections/platforms.

kunalspathak committed:
review feedback

kunalspathak committed:
fix the superpmi script

SingleAccretion commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -7182,17 +7283,32 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                     {
                         assert(value.Node() != tree);
 
+                        if (!isCommaTree || IsHoistingOverExcepSibling(value.Node(), hasExcep))
```
@kunalspathak could you show an example of how this safety check works? It is not obvious from the code why we treat `COMMA`s here specially.

kunalspathak replied:
The `isCommaTree` check is not really needed since we only set `hasExcep` if `isCommaTree == true`. I have also renamed the method to `IsHoistableOverExcepSibling` to make it more clear.

The idea is to not move `[000332]` before `[000328]` because if it is out of bounds, we would see the exception throw earlier or AV as in https://github.com/dotnet/runtime/pull/68061#issuecomment-1104572027.


```
               [000338] ---XG------                         *  COMMA     void  
     ( 12, 15) [000327] ---XG--H---                         +--*  COMMA     ref    <l:$258, c:$257>
     (  8, 11) [000328] ---X---H---                         |  +--*  BOUNDS_CHECK_Rng void   $245
     (  1,  1) [000329] -------H---                         |  |  +--*  CNS_INT   int    0 $c0
     (  3,  3) [000330] ---X---H---                         |  |  \--*  ARR_LENGTH int    $1c2
     (  1,  1) [000331] -------H---                         |  |     \--*  LCL_VAR   ref    V00 arg0         u:1 $100
     (  4,  4) [000332] n---G--H---                         |  \--*  IND       ref    <l:$256, c:$106>
     (  1,  2) [000333] -------H---                         |     \--*  ARR_ADDR  byref ref[] $82
     (  1,  2) [000334] -------N---                         |        \--*  ADD       byref  $2c0
     (  1,  1) [000335] -------H---                         |           +--*  LCL_VAR   ref    V00 arg0         u:1 $100
     (  1,  1) [000336] -------H---                         |           \--*  CNS_INT   long   16 $280
               [000337] -----------                         \--*  NOP       void  
```

SingleAccretion replied:
Out-of-order hoisting, I see. If we are hosting trees, we must hoist all of the side-effecting ones (in proper execution order), or only hoist the ones without side effects (which I suppose what the logic intends to do).

I still have some questions:
1) Why do we presume this problem can arise for `COMMA`s only?
2) Why is the logic gated on visiting the current tree? The comment above mentions:
```
At this point the stack contains (in top to bottom order):
  ...
  - ancestors of the current node and some of their descendants

  The ancestors have not been visited yet in post order so they're not hoistable
  (and they cannot become hoistable because the current node is not) but some of
  their descendants may have already been traversed and be hoistable.
```
So if some of those "descendants of ancestors" are a mix of such hoistable-not-hoistable nodes (like `[000328]` vs `[000332]` that we have above), the same problem could, presumably, arise.

Finally, as a remark, I have noticed there is quite a devious problem in the tree above, and it is the fact there is not enough information in it to determine that there is interference: `[000332]` _implicitly_ depends on being placed after the range check (it is why it is marked non-faulting). The only thing that saves us there is the fact we have `GTF_GLOB_REF` on it and the interference check is very conservative (`GLOB_REF` by itself doesn't imply any side effects, it just means the tree's value depends on the memory state, but VN already took care of that "in theory"...).

kunalspathak replied:
> Why do we presume this problem can arise for COMMAs only?

The problem does exist today as well. Today, we don't even check if we are hoisting past a tree that has side-effects (the example that you are stating). To me, `GT_COMMA` was the obvious one because of bounds check, but certainly if there are more examples where things can go wrong, we should address them here or in future PRs. 

> Why is the logic gated on visiting the current tree? 

The same reason. We visit the stack in execution order (bottom to top, as mentioned in comments) and I wanted to make sure to not hoist children past its parents, if parents have side-effects. Let me know if you can think of a better way to validate this.

> the same problem could, presumably, arise.

I noticed last week too while (in `main`) debugging something related where we hoist the checks ahead of side-effects but in that example, the range checks were getting eliminated by assertion prop and hence wouldn't have seen any functional difference.

SingleAccretion replied:
Hmm, so these are essentially "quirks" ("known holes in the algorithm"). They need to be documented as such.

> Let me know if you can think of a better way to validate this.

I would expect the "general" (conservative) logic to be something like:
```cs
bool canHoistExceptions = true; // We can hoist them as long as we haven't left any behind.

for (tree : stack)
{
    bool hoisted = false;

    if (tree->IsHoistable() && (canHoistExceptions || ((tree->gtFlags & GTF_ALL_EFFECT) == 0)))
    {
        hoisted = optHoistCandidate(...); // Can fail for budget reasons, etc
    }

    if (!hoisted && ((tree->gtFlags & GTF_EXCEPT) != 0))
    {
        canHoistExceptions = false;
    }
}
```
Not sure into how many regressions that would run into.

kunalspathak replied:
> Not sure how many regressions that would run into.

There are lot and it touches lot of methods that contains single-loops too that this PR don't intend to touch and I am not even sure if there are functional issues in these. I will leave it for future PRs.

Benchmarks windows/x64:
`363 total methods with Code Size differences (295 improved, 68 regressed), 58 unchanged.`

SingleAccretion replied:
Yeah, cannot say I didn't suspect something like that would turn up :). My only remaining request here is for comments then.

kunalspathak committed:
Revert superpmi asmdiffs change

kunalspathak committed:
Rename method

kunalspathak committed:
Add a comment
